fairlearn/preprocessing/_information_filter.py

```diff
@@ -0,0 +1,80 @@
+# Copyright (c) Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils import check_array
+from sklearn.utils.validation import check_is_fitted
+
+
+class InformationFilter(BaseEstimator, TransformerMixin):
+    r"""
+    A component that filters out sensitive correlations in a dataset.
+
+    The `InformationFilter` uses a variant of the Gram-Schmidt process
+    to filter information out of the dataset. This can be useful if you
+    want to filter information out of a dataset because of fairness.
+    To explain how it works: given a training matrix :math:`X` that contains
+    columns :math:`x_1, ..., x_k`. If we assume columns :math:`x_1` and :math:`x_2`
+    to be the sensitive columns then the information-filter will
+    remove information by applying these transformations;
+
+    .. math::
+
+       \\begin{split}
+       v_1 & = x_1 \\\\
+       v_2 & = x_2 - \\frac{x_2 v_1}{v_1 v_1}\\\\
+       v_3 & = x_3 - \\frac{x_k v_1}{v_1 v_1} - \\frac{x_2 v_2}{v_2 v_2}\\\\
+       ... \\\\
+       v_k & = x_k - \\frac{x_k v_1}{v_1 v_1} - \\frac{x_2 v_2}{v_2 v_2}
+       \\end{split}
+
+    Concatenating our vectors (but removing the sensitive ones) gives us
+    a new training matrix :math:`X_{filtered} =  [v_3, ..., v_k]`. The final output
+    is an interpolation between the original dataset (without sensitive columns)
+    and the fitlered dataset.
+
+    .. math::
+
+      X_{\text{tfm}} = \alpha X_{\text{filtered}} + (1-\alpha) X_{\text{orig}}
+
+    :param columns: list of columns to filter out this can be a sequence of either int
+                    (in the case of numpy) or string (in the case of pandas).
+    :param alpha: parameter to control how much to filter, for alpha=1 we filter out
+                  all information while for alpha=0 we don't apply any.
+    :param center: setting to tell if this preprocessing step should center the data for
+                   numerical stability
+    """
+
+    def __init__(self, columns, alpha=1, center=True):
+        self.columns = columns
+        self.alpha = alpha
+        self.center = center
+
+    def _split_X(self, X):
+        """Split up X into a sensitive and non-sensitive group."""
+        sensitive = self.columns
+        non_sensitive = [i for i in range(X.shape[1]) if i not in sensitive]
```

Comment:
Have you tested this when `self.columns` is a list of strings? In that case it looks like you retain all of the original columns.

Reply:
Right, we should definitely check that `X` actually had those columns. Otherwise users may think they correctly applied it but they didn't.

Reply:
The current implementation I've made assumes a numpy array which has numeric column names. I didn't know what standard you folks use if it is a pandas dataframe. Do the fairlearn components assume that the input can be either `np.ndarray` or `pd.DataFrame`?

Reply:
X can be ndarray or DataFrame, yes. The API doc (hopefully) says that in every mitigation technique.

We don't do a lot with X itself since we just pass it on to the underlying estimator so far, so we haven't really hit this problem a lot.

It's only two types, so I think it's fine to do this with an `if` and check whether the columns exist. What do you think?